### PR TITLE
FIX: Added condition for list and set in build opts

### PR DESCRIPTION
**Pull request by David-Rod**
PR adds checks to ensure that we are able to handle `ScopeRecord` instances within a list or set.

When adding a usage example to `q2-feature-table`, it was discovered that the framework handled lists and `ScopeRecord` types, but could not take in lists or sets with ScopeRecord types as values within that list/set.

This PR adds two conditions that update the `value` variable, using set or list comprehension (depending on the case), to handle such cases.

@thermokarst Is there anything you would recommend for testing these changes moving forward?

Thanks in advance.

### CODE REVIEW

**Comment by thermokarst** [comment]
Hey @David-Rod - thanks for this PR, and thanks for filling me in on the details!

## Dummy Plugin Usage Examples

For testing, you will need to create a few new usage examples in the dummy plugin, drop those here:

https://github.com/qiime2/qiime2/blob/master/qiime2/core/testing/examples.py

Don't forget to register those examples in the dummy plugin (like you did with the examples you made for q2-feature-table).

Looks like this method might let you address `Set` and `List` in one shot:

https://github.com/qiime2/qiime2/blob/f89eaadf693c85874dc9fae4017a22b5ddcf7ed0/qiime2/core/testing/plugin.py#L437-L463

## Unit Tests

Once you have some usage examples in place, you can write some unit tests:

https://github.com/qiime2/qiime2/blob/f89eaadf693c85874dc9fae4017a22b5ddcf7ed0/qiime2/sdk/tests/test_usage.py#L28

You are mostly just wanting to run your usage example(s), with the `DiagnosticUsage` driver (see the other tests in that suite) and make sure that everything looks good. Think about your q2-feature-table examples while doing this - keep in mind the things that prompted you to write this PR in the first place - those are good things to check for!

Also, please drop in a test or two in this section:

https://github.com/qiime2/qiime2/blob/f89eaadf693c85874dc9fae4017a22b5ddcf7ed0/qiime2/sdk/tests/test_usage.py#L191

You won't need any usage examples there, but basically you should add two tests like https://github.com/qiime2/qiime2/blob/f89eaadf693c85874dc9fae4017a22b5ddcf7ed0/qiime2/sdk/tests/test_usage.py#L205, the first should check for a list, the other for a set.

## Misc

Other miscellaneous things: I don't think any changes need to be made to the other Usage drivers in this repo, but it might be worth duplicating your test that you write for `DiagnosticUsage`:

https://github.com/qiime2/qiime2/blob/f89eaadf693c85874dc9fae4017a22b5ddcf7ed0/qiime2/sdk/tests/test_usage.py#L28

in `ArtifactAPIUsage`, just to cover our bases:

https://github.com/qiime2/qiime2/blob/f89eaadf693c85874dc9fae4017a22b5ddcf7ed0/qiime2/tests/test_artifact_api.py#L122

## Summary

1. Create some new usage examples in the dummy plugin
2. Add some unit tests that exercise your examples with `DiagnosticUsage`
3. Add some unit tests that verify the basic functionality you added to `UsageInputs` (no need to use your usage examples)
4. Duplicate tests in step 2 for `ArtifactAPIUsage`

Please let me know if you need clarification or have any questions! I am happy to help out, just let me know. Thanks!

File: qiime2/sdk/usage.py [hunk 1]
```diff
@@ -88,6 +88,14 @@ def build_opts(self, signature, scope):
                 if isinstance(self.values[name], ScopeRecord) \
                         and self.values[name].ref in scope.records:
                     value = self.values[name].result
+                elif (isinstance(self.values[name], list)
+                      and self.values[name]
+                      and isinstance(self.values[name][0], ScopeRecord)):
```
Comment (thermokarst) [inline qiime2/sdk/usage.py]: Perhaps for a "belt-and-suspenders" approach here, we should consider checking that _all_ of the records in the collection are `ScopeRecord` instances. The same consideration applies to the `set` branch below.

Comment (David-Rod) [reply qiime2/sdk/usage.py]: Would something like the following work to check the records in the collections mentioned above:

`all(isinstance(n, ScopeRecord) for n in self.values[name]))`

I've tested this solution for both set and list types that contain string values, and they both seem to work.

Comment (thermokarst) [reply qiime2/sdk/usage.py]: Works for me!

**Comment by David-Rod** [comment]
Hey @thermokarst,

I believe I was able to accomplish the items you listed above. I am slightly unsure about the tests that I created, but I think most of what you may be looking for is here. 

Thanks in advance for your review!

**Comment by thermokarst** [comment]
Thanks @David-Rod - will take a look soon!

File: qiime2/core/testing/examples.py [hunk 2]
```diff
@@ -50,6 +50,26 @@ def mdc1_factory():
                                                     name='id')))
 
 
+def intSequence_list_factory():
```
Comment (thermokarst) [inline qiime2/core/testing/examples.py:53]: Please reformat this as a [PEP 8-compliant name](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)

```suggestion
def int_sequence_list_factory():
```

File: qiime2/core/testing/examples.py [hunk 3]
```diff
@@ -50,6 +50,26 @@ def mdc1_factory():
...
+    int1 = ints1_factory()
+    int2 = ints2_factory()
+    return [int1, int2]
+
+
+def singleInt1_factory():
```
Comment (thermokarst) [inline qiime2/core/testing/examples.py:59]: PEP 8 (see above)

File: qiime2/core/testing/examples.py [hunk 4]
```diff
@@ -50,6 +50,26 @@ def mdc1_factory():
...
+    return Artifact.import_data(SingleInt, 10)
+
+
+def singleInt2_factory():
```
Comment (thermokarst) [inline qiime2/core/testing/examples.py:63]: PEP 8 (see above)

File: qiime2/core/testing/examples.py [hunk 5]
```diff
@@ -50,6 +50,26 @@ def mdc1_factory():
...
+    return Artifact.import_data(SingleInt, 11)
+
+
+def singleInt1_set_factory():
```
Comment (thermokarst) [inline qiime2/core/testing/examples.py:67]: PEP 8 (see above)

File: qiime2/sdk/tests/test_usage.py [hunk 6]
```diff
@@ -231,6 +248,13 @@ def test_validated_missing_optional_value(self):
         ui.validate(self.signature)
         self.assertTrue(True)
 
+    def test_type_of_input(self):
+        test_inputs = usage.UsageInputs(x=[1, 2, 3], z={7, 8, 9})
+        test_inputs.validate(self.signature)
+
+        self.assertEqual(list, type(test_inputs.values['x']))
+        self.assertEqual(set, type(test_inputs.values['z']))
```
Comment (thermokarst) [inline qiime2/sdk/tests/test_usage.py:256]: Great! A slightly more idiomatic approach would use https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIsInstance

```suggestion
        self.assertIsInstance(test_inputs.values['x'], list)
        self.assertIsInstance(test_inputs.values['z'], set)
```

File: qiime2/tests/test_artifact_api.py [hunk 7]
```diff
@@ -252,6 +252,22 @@ def test_metadata_column_from_factory(self):
...
+    def test_use_merge_feature_table(self):
+        action = self.plugin.actions['variadic_input_method']
+        use = ArtifactAPIUsage()
+        action.examples['feature_table_merge_example'](use)
+
+        exp = """\
+from qiime2.plugins.dummy_plugin.methods import variadic_input_method
+
+out, = variadic_input_method(
+    ints=int,
```
Comment (thermokarst) [inline qiime2/tests/test_artifact_api.py:267]: Note to @thermokarst - this driver should template in the factory body to make this more clear.

**Comment by David-Rod** [comment]
Regarding the changes for the latest commit, I made a slight change to the validation for inputs that are provided to accommodate the changes made in https://github.com/qiime2/q2-feature-table/pull/226. I know this change does not capture all possibilities when we are trying to provide more than one input, but it may be a step in the right direction. Hopefully your review will provide some more clarity. Thanks.

File: qiime2/sdk/usage.py [hunk 8]
```diff
@@ -77,8 +77,10 @@ def validate(self, signature):
 
         extra = provided - exp_inputs - exp_params
         if len(extra) > 0:
-            raise ValueError('Extra input(s) or parameter(s): %r' %
-                             (extra, ))
+            for provided_value in extra:
```
Comment (thermokarst) [inline qiime2/sdk/usage.py:80]: @David-Rod - I suspect you won't need any of this once https://github.com/qiime2/qiime2/pull/530 is merged.

**Comment by thermokarst** [comment]
This is blocked by #530

**Comment by thermokarst** [comment]
Hi @David-Rod, this PR should be unblocked, but looks like you will need to merge in the latest master and deal with the merge conflicts. Please let me know if you need a hand with that, I am happy to help.

**Comment by thermokarst** [comment]
Hey @David-Rod - don't worry about the merge conflicts - I have some free time and can handle it, stay tuned.

**Pull request merged**